### PR TITLE
Get rest of Lib/Parser building

### PR DESCRIPTION
**Pull request by digitalinfinity**
Compile the following files:
    - CharClassifier.cpp
    - CharSet.cpp
    - Parse.cpp
    - RegexParser.cpp
    - RegexPattern.cpp
    - RegexRuntime.cpp
    - Scan.cpp
    - StandardChars.cpp
    - TextbookBoyerMoore.cpp

Get scan.cpp compiling

Couple noteworthy items:
    - added _ltow_s/_ltow to the PAL
    - Exposed ULongMult from the PAL
    - Switched a __try/__finally in Scan.cpp to TryFinally

### CODE REVIEW

**Comment by digitalinfinity** [comment]
@Jianchun can you review please?

File: lib/Parser/Parse.cpp [hunk 1]
```diff
@@ -10868,7 +10868,7 @@ ParseNodePtr Parser::Parse(LPCUTF8 pszSrc, size_t offset, size_t length, charcou
 
     m_scriptContext->AddSourceSize(m_length);
 
-    if(!m_parseType != ParseType_Deferred)
+    if (m_parseType != ParseType_Deferred)
```
Comment (jianchun) [inline lib/Parser/Parse.cpp]: Does this fix a bug?

File: lib/Parser/RegexParser.cpp [hunk 2]
```diff
@@ -2948,7 +2949,7 @@ namespace UnifiedRegex
     {
         Assert(program->source == 0);
 
-        program->source = _u("");
+        program->source = const_cast<Char*>(_u(""));
```
Comment (jianchun) [inline lib/Parser/RegexParser.cpp]: Can "program->source" be made const Char*?

File: lib/Parser/StandardChars.h [hunk 3]
```diff
@@ -225,6 +225,8 @@ namespace UnifiedRegex
             return true;
         }
 
+        TrivialCaseMapper() {}
+
```
Comment (jianchun) [inline lib/Parser/StandardChars.h]: nit pick: trailing spaces. What makes this constructor needed... the CompileAssert?

File: pal/inc/rt/palrt.h [hunk 4]
```diff
@@ -1076,6 +1076,33 @@ errno_t __cdecl _itow_s(int _Value, WCHAR *_Dst, size_t _SizeInWords, int _Radix
...
+errno_t __cdecl _ltow_s(long _Value, WCHAR (&_Dst)[_SizeInWords], int _Radix)
+{
+    return _ltow_s(_Value, _Dst, _SizeInWords, _Radix);
+}
+#endif
+
+#if _SAFECRT_USE_INLINES
+
+__inline
+errno_t __cdecl _ltow_s(long _Value, WCHAR *_Dst, size_t _SizeInWords, int _Radix)
```
Comment (jianchun) [inline pal/inc/rt/palrt.h]: We will also need a bunch other forms... ultow_s, ui64tow_s...

File: lib/Parser/Parse.cpp (hunk 1 above)
Comment (digitalinfinity) [inline lib/Parser/Parse.cpp]: Yeah, I think so- I think that was a typo there (we check for m_parseType != ParseType_Deferred when we fire the start event)

File: lib/Parser/RegexParser.cpp (hunk 2 above)
Comment (digitalinfinity) [inline lib/Parser/RegexParser.cpp]: That was the first thing I tried but quite a bit of code assumes its not const so went with the cheaper fix of maintaining the current assumption

File: pal/inc/rt/palrt.h (hunk 4 above)
Comment (digitalinfinity) [inline pal/inc/rt/palrt.h]: ui64 already exists I think- ultow_s, we can add- I vote for adding them as and when we need it

**Comment by jianchun** [comment]
LGTM! Thanks for completing these!

File: lib/Parser/StandardChars.h (hunk 3 above)
Comment (digitalinfinity) [inline lib/Parser/StandardChars.h]: No, we have a const instance of TrivialCaseMapper. C++ requires that any const objects that are default constructed should have a default constructor- see http://stackoverflow.com/questions/7411515/why-does-c-require-a-user-provided-default-constructor-to-default-construct-a
I can add a comment

Comment (jianchun) [inline lib/Parser/StandardChars.h]: Good to know! Thanks!

File: lib/Parser/Parse.cpp (hunk 1 above)
Comment (digitalinfinity) [inline lib/Parser/Parse.cpp]: Confirmed with @abchatra that this is an existing bug

Comment (jianchun) [inline lib/Parser/Parse.cpp]: Agree, it should be matching line 10681.

**Pull request merged**